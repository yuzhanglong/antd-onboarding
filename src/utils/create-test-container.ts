import { ReactElement } from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';

export const createTestContainer = (cmp: ReactElement) => {
  let container: ReactTestRenderer = null;
  act(() => {
    container = create(cmp);
  });

  return container;
};
