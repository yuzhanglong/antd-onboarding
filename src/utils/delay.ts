export const delay = (timeout: number, cb: () => any) => {
  return new Promise((resolve) => {
    setTimeout(() => {
      cb();
      resolve(true);
    }, timeout);
  });
};
