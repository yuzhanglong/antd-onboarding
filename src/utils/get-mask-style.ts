/**
 * 获取遮罩样式
 *
 * @author yuzhanglong
 * @date 2021-09-14 18:07:12
 * @param element 遮罩的焦点元素，在视觉上体现为这个元素所在的部分可见，其它的部分都会被遮罩遮挡
 * @param container 遮罩的容器，即遮罩的作用范围
 */
export const getMaskStyle = (element: HTMLElement, container: HTMLElement) => {
  if (!element) {
    return {};
  }

  const {
    scrollHeight: containerScrollHeight,
    scrollWidth: containerScrollWidth,
    scrollTop: containerScrollTop,
    scrollLeft: containerScrollLeft
  } = container;

  const {
    height: elementHeight,
    width: elementWidth,
    left: elementLeft,
    top: elementTop
  } = element.getBoundingClientRect();

  const elementTopWithScroll = containerScrollTop + elementTop;
  const elementLeftWithScroll = containerScrollLeft + elementLeft;

  return {
    width: containerScrollWidth,
    height: containerScrollHeight,
    borderTopWidth: Math.max(elementTopWithScroll, 0),
    borderLeftWidth: Math.max(elementLeftWithScroll, 0),
    borderBottomWidth: Math.max(containerScrollHeight - elementHeight - elementTopWithScroll, 0),
    borderRightWidth: Math.max(containerScrollWidth - elementWidth - elementLeftWithScroll, 0)
  };
};
