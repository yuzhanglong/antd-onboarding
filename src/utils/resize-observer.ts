import { MaskStyleCheckObserver } from '../types';

export const resizeObserver: MaskStyleCheckObserver = (element, check) => {
  let timer = null;

  const onWindowResize = () => {
    if (timer) {
      window.cancelAnimationFrame(timer);
    }

    timer = window.requestAnimationFrame(() => {
      check();
    });
  };


  return {
    observe: () => {
      window.addEventListener('resize', onWindowResize);
    },
    destroy: () => {
      window.removeEventListener('resize', onWindowResize);
    }
  };
};
