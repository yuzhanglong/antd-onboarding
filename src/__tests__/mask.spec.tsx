import React from 'react';
import { Mask } from '../';
import { noop } from 'lodash';
import { createTestContainer } from '../utils/create-test-container';
import { MASK_ANIMATION_TIME } from '../';
import { act } from '@testing-library/react';

jest.useFakeTimers();

describe('引导组件遮罩层相关测试', () => {
  const testBaseElement = document.createElement('div');
  testBaseElement.style.height = '200px';
  testBaseElement.style.width = '200px';
  document.body.appendChild(testBaseElement);

  test('mask 配置的监听器应该按预期销毁', () => {
    const destroyCallback = jest.fn();
    const container = createTestContainer(
      <Mask
        styleChecker={
          () => {
            return {
              observe: noop,
              destroy: destroyCallback
            };
          }
        }
        element={testBaseElement} />
    );

    // 未销毁
    expect(destroyCallback).toBeCalledTimes(0);

    act(() => {
      container.unmount();
    });

    expect(destroyCallback).toBeCalledTimes(1);
  });

  test('基准元素为空, mask 不会去计算样式或者监听基准元素, 仅渲染一个无样式的空 div', () => {
    const checkCallback = jest.fn();
    const Wrapper = () => {
      return (
        <Mask element={null} styleChecker={() => {
          return {
            observe: checkCallback,
            destroy: noop
          };
        }} />
      );
    };

    const container = createTestContainer(<Wrapper />);


    expect(checkCallback).toBeCalledTimes(0);
    expect(container.baseElement).toMatchSnapshot();
  });

  test('配置了基准元素, mask 样式应该被计算后写入 dom', async () => {
    const container = createTestContainer(<Mask element={testBaseElement} />);
    expect(container.baseElement).toMatchSnapshot();
  });

  test('测试遮罩层可见能力, 默认为 true', () => {
    const container = createTestContainer(
      <Mask
        element={testBaseElement}
        visible={false} />
    );


    expect(container.baseElement).toMatchSnapshot();

    act(() => {
      container.rerender(
        <Mask
          element={testBaseElement}
          visible={true} />
      );
    });

    expect(container.baseElement).toMatchSnapshot();
  });

  test('测试 mask 的两个生命周期 onAnimationStart / onAnimationEnd', () => {
    const animationStartCallback = jest.fn();
    const animationEndCallback = jest.fn();
    const container = createTestContainer(
      <Mask
        onAnimationStart={animationStartCallback}
        onAnimationEnd={animationEndCallback}
        element={testBaseElement} />
    );

    expect(animationStartCallback).toBeCalledTimes(1);
    expect(animationEndCallback).toBeCalledTimes(0);

    // 在动画结束的时间后组件应该调用 animationEndCallback
    jest.advanceTimersByTime(MASK_ANIMATION_TIME);
    expect(animationStartCallback).toBeCalledTimes(1);
    expect(animationEndCallback).toBeCalledTimes(1);


    // 在 element 更新后重复上述生命周期
    const el2 = document.createElement('div');
    act(() => {
      container.rerender(
        <Mask
          onAnimationStart={animationStartCallback}
          onAnimationEnd={animationEndCallback}
          element={el2} />
      );
    });

    expect(animationStartCallback).toBeCalledTimes(2);
    expect(animationEndCallback).toBeCalledTimes(1);
    jest.advanceTimersByTime(MASK_ANIMATION_TIME);
    expect(animationStartCallback).toBeCalledTimes(2);
    expect(animationEndCallback).toBeCalledTimes(2);
  });

  test('测试 mask 子组件的写入', () => {
    const container = createTestContainer(
      <Mask
        element={testBaseElement}
        renderMaskContent={(wrapper) => {
          return (
            <div>
              {wrapper}
              hello world!
            </div>
          );
        }} />
    );
    expect(container.baseElement).toMatchSnapshot();
  });
});
