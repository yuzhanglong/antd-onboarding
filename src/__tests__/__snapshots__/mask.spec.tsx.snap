// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`引导组件遮罩层相关测试 基准元素为空, mask 不会去计算样式或者监听基准元素, 仅渲染一个无样式的空 div 1`] = `
<div
  className="mask mask-animation"
  style={Object {}}
/>
`;

exports[`引导组件遮罩层相关测试 测试 mask 子组件的写入 1`] = `
<div
  className="mask mask-animation"
  style={
    Object {
      "borderBottomWidth": 0,
      "borderLeftWidth": 1000,
      "borderRightWidth": 0,
      "borderTopWidth": 1000,
      "height": 1000,
      "width": 1000,
    }
  }
>
  <div>
    <div
      className="mask-content"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
    hello world!
  </div>
</div>
`;

exports[`引导组件遮罩层相关测试 测试遮罩层可见能力, 默认为 true 1`] = `
<div
  className="mask mask-animation mask-not-visible"
  style={
    Object {
      "borderBottomWidth": 0,
      "borderLeftWidth": 1000,
      "borderRightWidth": 0,
      "borderTopWidth": 1000,
      "height": 1000,
      "width": 1000,
    }
  }
/>
`;

exports[`引导组件遮罩层相关测试 测试遮罩层可见能力, 默认为 true 2`] = `
<div
  className="mask mask-animation"
  style={
    Object {
      "borderBottomWidth": 0,
      "borderLeftWidth": 1000,
      "borderRightWidth": 0,
      "borderTopWidth": 1000,
      "height": 1000,
      "width": 1000,
    }
  }
/>
`;

exports[`引导组件遮罩层相关测试 配置了基准元素, mask 样式应该被计算后写入 dom 1`] = `
<div
  className="mask mask-animation"
  style={
    Object {
      "borderBottomWidth": 0,
      "borderLeftWidth": 1000,
      "borderRightWidth": 0,
      "borderTopWidth": 1000,
      "height": 1000,
      "width": 1000,
    }
  }
/>
`;
